File: judo-data/src/main/java/com/judopay/arch/api/ApiHeadersInterceptor.java
package com.judopay.arch.api;

import android.os.Build;

import com.judopay.rest.BuildConfig;
import com.squareup.okhttp.Headers;
import com.squareup.okhttp.Interceptor;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;

import java.io.IOException;
import java.util.HashMap;
import java.util.Locale;

class ApiHeadersInterceptor implements Interceptor {

    private static final String AUTHORIZATION_HEADER = "Authorization";
    private static final String CONTENT_TYPE_HEADER = "Content-Type";
    private static final String ACCEPT_HEADER = "Accept";
    private static final String API_VERSION_HEADER = "Api-Version";
    private static final String CACHE_CONTROL_HEADER = "Cache-Control";

    private static final String JSON_MIME_TYPE = "application/json";
    private static final String API_VERSION = "4.1.0";
    private static final String CACHE_CONTROL = "no-cache";
    private static final String SDK_VERSION_HEADER = "Sdk-Version";
    private static final String USER_AGENT_HEADER = "User-Agent";

    private final UserAgent userAgent;
    private final AuthorizationEncoder authorizationEncoder;

    public ApiHeadersInterceptor(AuthorizationEncoder authorizationEncoder) {
        this.authorizationEncoder = authorizationEncoder;
        this.userAgent = new UserAgent(BuildConfig.VERSION_NAME, Build.VERSION.RELEASE,
                Build.MANUFACTURER, Build.MODEL, Locale.getDefault().getDisplayName());
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request.Builder builder = chain.request()
                .newBuilder()
                .headers(getHeaders());

        return chain.proceed(builder.build());
    }

    private Headers getHeaders() {
        HashMap<String, String> headers = new HashMap<>();

        headers.put(AUTHORIZATION_HEADER, authorizationEncoder.getAuthorization());
        headers.put(CONTENT_TYPE_HEADER, JSON_MIME_TYPE);
        headers.put(ACCEPT_HEADER, JSON_MIME_TYPE);
        headers.put(API_VERSION_HEADER, API_VERSION);
        headers.put(CACHE_CONTROL_HEADER, CACHE_CONTROL);
        headers.put(SDK_VERSION_HEADER, BuildConfig.VERSION_NAME);
        headers.put(USER_AGENT_HEADER, userAgent.toString());

        return Headers.of(headers);
    }
}

File: judokit-android/src/main/java/com/judopay/judokit/android/api/interceptor/ApiHeadersInterceptor.kt
package com.judopay.judokit.android.api.interceptor

import com.judopay.judokit.android.BuildConfig
import com.judopay.judokit.android.api.AppMetaDataProvider
import com.judopay.judokit.android.api.model.Authorization
import okhttp3.Headers
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

private const val CONTENT_TYPE_HEADER = "Content-Type"
private const val ACCEPT_HEADER = "Accept"
private const val API_VERSION_HEADER = "Api-Version"
private const val CACHE_CONTROL_HEADER = "Cache-Control"
private const val UI_MODE_HEADER = "UI-Client-Mode"
private const val SDK_VERSION_HEADER = "Sdk-Version"
private const val USER_AGENT_HEADER = "User-Agent"
private const val JSON_MIME_TYPE = "application/json"
private const val API_VERSION = "5.6.0"
private const val BANK_API_VERSION = "2.0.0.0"
private const val CACHE_CONTROL = "no-cache"
private const val CUSTOM_UI_MODE = "Custom-UI"

private const val BANK_ENDPOINT = "/order/bank"

internal class ApiHeadersInterceptor(
    private val authorization: Authorization,
    private val appMetaDataProvider: AppMetaDataProvider
) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val path = chain.request().url.encodedPath
        val isSaleRequest = path.startsWith(BANK_ENDPOINT)
        val headers = getHeaders(isSaleRequest)
        val request = chain.request()
            .newBuilder()
            .headers(headers)
            .build()

        return chain.proceed(request)
    }

    private fun getHeaders(isSaleRequest: Boolean) = Headers.Builder()
        .addAll(authorization.headers)
        .add(CONTENT_TYPE_HEADER, JSON_MIME_TYPE)
        .add(ACCEPT_HEADER, JSON_MIME_TYPE)
        .add(API_VERSION_HEADER, if (isSaleRequest) BANK_API_VERSION else API_VERSION)
        .add(CACHE_CONTROL_HEADER, CACHE_CONTROL)
        .add(SDK_VERSION_HEADER, "Android-${BuildConfig.VERSION_NAME}")
        .addUnsafeNonAscii(USER_AGENT_HEADER, appMetaDataProvider.userAgent)
        .add(UI_MODE_HEADER, CUSTOM_UI_MODE)
        .build()
}
